### PR TITLE
Report source incompatibilities caused by bumped dependencies

### LINKED ISSUE #87: versionPolicyReportDependencyIssues does not report source incompatibilities
It only reports binary incompatibilities.

### PR DESCRIPTION
**Pull request by julienrf**
Fixes #87

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -11,7 +11,7 @@ inThisBuild(List(
       url("https://github.com/alexarchambault")
     )
   ),
-  versionPolicyIntention := Compatibility.BinaryAndSourceCompatible
+  versionPolicyIntention := Compatibility.BinaryCompatible
```
Comment (julienrf) [inline build.sbt:14]: Next release will be 1.1.0 :relaxed:

File: sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala [hunk 2]
```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
...
+    apply(
+      Compatibility.BinaryCompatible,
+      currentModules,
+      previousModules,
+      reconciliations,
+      defaultReconciliation
+    )
+
+  private[sbtversionpolicy] def apply(
+    compatibilityIntention: Compatibility,
```
Comment (julienrf) [inline sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala:81]: We now take into account the compatibility level to check (none, binary compatible, or binary and source compatible)

File: sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala [hunk 3]
```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
...
-
+  @deprecated("This method is internal")
   def apply(
     currentModules: Map[(String, String), String],
     previousModules: Map[(String, String), String],
     reconciliations: Seq[(ModuleMatchers, VersionCompatibility)],
     defaultReconciliation: VersionCompatibility
+  ): DependencyCheckReport =
+    apply(
+      Compatibility.BinaryCompatible,
```
Comment (julienrf) [inline sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala:73]: Previous behavior was to only check binary compatibility.

File: sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala [hunk 4]
```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
...
             .getOrElse(defaultReconciliation)
-          if (reconciliation.isCompatible(ver, currentVersion))
-            CompatibleVersion(currentVersion, ver, reconciliation)
+          val isCompatible =
+            if (compatibilityIntention == Compatibility.BinaryAndSourceCompatible) {
+              isBinaryCompatible(currentVersion, previousVersion, reconciliation) &&
+                isSourceCompatible(currentVersion, previousVersion, reconciliation)
+            } else {
+              isBinaryCompatible(currentVersion, previousVersion, reconciliation)
+            }
```
Comment (julienrf) [inline sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala:135]: According to the intended compatibility, we check the source compatibility or not.

File: sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala [hunk 5]
```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
...
+          else
+            currentMinor == previousMinor && currentPatch >= previousPatch
+        }
+      case VersionCompatibility.SemVerSpec =>
+        val (currentMajor, currentMinor, currentPatch) = extractSemVerNumbers(currentVersion)
+        val (previousMajor, previousMinor, previousPatch) = extractSemVerNumbers(previousVersion)
+        currentMajor == previousMajor &&
+          currentMinor == previousMinor &&
+          currentPatch >= previousPatch
     }
```
Comment (julienrf) [inline sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala:170]: We implement source compatibility checking here, according to the versioning scheme of the dependency.

File: sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala [hunk 6]
```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
...
+
+  private def extractSemVerNumbers(versionString: String): (Int, Int, Int) = {
+    val version = Version(versionString)
+    if (!(version.items.size == 3 && version.items.forall(_.isInstanceOf[Version.Number]))) {
+      throw new MessageOnlyException(s"Version ${versionString} is not a semantic version. Set the correct versioning scheme via versionPolicyDependencySchemes.")
```
Comment (sjrd) [inline sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala:175]: Isn't that going to blow up if we depend on an M/RC of a library, even temporarily? Is this the desired behavior?

Comment (julienrf) [reply sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala:170]: > Isn't that going to blow up if we depend on an M/RC of a library, even temporarily?

Yes.

> Is this the desired behavior?

Yes, but now that I’m thinking more about it, I have a better idea. Let me push it.

**Review by sjrd — APPROVED** [review]
Indeed, that's better :)

**Pull request merged**